tool/tsh/db.go

```diff
@@ -858,54 +858,132 @@ func newDatabaseInfo(cf *CLIConf, tc *client.TeleportClient, route tlsca.RouteTo
+		dbInfo.Username = tc.Username
+	}
+	// recheck to see if we can avoid fetching the roleset to set defaults.
+	needDBUser := dbInfo.Username == "" && role.RequireDatabaseUserMatcher(dbInfo.Protocol)
+	needDBName := dbInfo.Database == "" && role.RequireDatabaseNameMatcher(dbInfo.Protocol)
+	if !needDBUser && !needDBName {
+		return &dbInfo, nil
+	}
+
+	profile, err := tc.ProfileStatus()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var proxy *client.ProxyClient
+	err = client.RetryWithRelogin(cf.Context, tc, func() error {
+		proxy, err = tc.ConnectToProxy(cf.Context)
+		return trace.Wrap(err)
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer proxy.Close()
+
+	roleSet, err := fetchRoleSetForCluster(cf.Context, profile, proxy, tc.SiteName)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if needDBUser {
+		dbUser, err := getDefaultDBUser(db, roleSet)
 		if err != nil {
 			return nil, trace.Wrap(err)
 		}
-		dbInfo.Protocol = db.GetProtocol()
+		log.Debugf("Defaulting to the allowed database user %q\n", dbUser)
+		dbInfo.Username = dbUser
 	}
-	if dbInfo.Username == "" {
-		db, err := dbInfo.GetDatabase(cf, tc)
+	if needDBName {
+		dbName, err := getDefaultDBName(db, roleSet)
 		if err != nil {
 			return nil, trace.Wrap(err)
 		}
-		// If database has admin user defined, we're most likely using automatic
-		// user provisioning so default to Teleport username unless database
-		// username was provided explicitly.
-		if db.GetAdminUser() != "" {
-			log.Debugf("Defaulting to Teleport username %q as database username.", tc.Username)
-			dbInfo.Username = tc.Username
-		} else {
-			switch dbInfo.Protocol {
-			// When generating certificate for MongoDB access, database username must
-			// be encoded into it. This is required to be able to tell which database
-			// user to authenticate the connection as Elasticsearch needs database username too.
-			case defaults.ProtocolMongoDB, defaults.ProtocolElasticsearch, defaults.ProtocolOracle, defaults.ProtocolOpenSearch:
-				return nil, trace.BadParameter("please provide the database user name using the --db-user flag")
-			case defaults.ProtocolRedis:
-				// Default to "default" in the same way as Redis does. We need the username to check access on our side.
-				// ref: https://redis.io/commands/auth
-				log.Debugf("Defaulting to Redis username %q as database username.", defaults.DefaultRedisUsername)
-				dbInfo.Username = defaults.DefaultRedisUsername
-			}
-		}
+		log.Debugf("Defaulting to the allowed database name %q\n", dbName)
+		dbInfo.Database = dbName
 	}
-	if dbInfo.Database != "" {
-		switch dbInfo.Protocol {
-		case defaults.ProtocolDynamoDB:
-			log.Warnf("Database %v protocol %v does not support --db-name flag, ignoring --db-name=%v",
-				dbInfo.ServiceName, defaults.ReadableDatabaseProtocol(dbInfo.Protocol), dbInfo.Database)
-			dbInfo.Database = ""
+
+	return &dbInfo, nil
```

Comment:
nit: For readability can we move this logic into a  sperate function  ?  Could you also add a comment describing  default user behavior ?

Reply:
I had auto-merge enabled so this merged after you approved it, but I can do a followup PR